core-common/src/main/java/org/glassfish/jersey/internal/inject/ParamConverters.java

```diff
@@ -305,6 +309,125 @@ public String toString(T value) throws IllegalArgumentException {
+     * by invoking {@link AggregatedProvider}.
+     */
+    @Singleton
+    public static class OptionalProvider implements ParamConverterProvider {
+
+        // Delegates to this provider when the type of Optional is extracted.
+        private final AggregatedProvider aggregated;
+
+        @Inject
+        public OptionalProvider(AggregatedProvider aggregated) {
+            this.aggregated = aggregated;
+        }
+
+        @Override
+        public <T> ParamConverter<T> getConverter(Class<T> rawType, Type genericType, Annotation[] annotations) {
+            final Optionals optionals = Optionals.getOptional(rawType);
+            return (optionals == null) ? null : new ParamConverter<T>() {
+
+                @Override
+                public T fromString(String value) {
+                    if (value == null) {
+                        return (T) optionals.empty();
+                    } else {
+                        if (genericType instanceof ParameterizedType) {
+                            // Optional is parameterized
+                            ParameterizedType parametrized = (ParameterizedType) genericType;
+                            Type type = parametrized.getActualTypeArguments()[0];
+                            T val = aggregated.getConverter((Class<T>) type, type, annotations).fromString(value.toString());
+                            if (val != null) {
+                                return (T) optionals.of(val);
+                            } else {
+                                /*
+                                 *  In this case we don't send Optional.empty() because 'value' is not null.
+                                 *  But we return null because the provider didn't find how to parse it.
+                                 */
+                                return null;
+                            }
+                        } else {
+                            // Others are not parameterized
+                            return (T) optionals.of(value);
+                        }
+                    }
+                }
+
+                @Override
+                public String toString(T value) throws IllegalArgumentException {
+                    /*
+                     *  Unfortunately 'orElse' cannot be stored in an Optional. As only one value can
+                     *  be stored, it makes no sense that 'value' is Optional. It can just be the value.
+                     *  We don't fail here but we don't process it.
+                     */
+                    return null;
+                }
+            };
+        }
+
+        private static enum Optionals {
+
+            OPTIONAL(Optional.class) {
+                @Override
+                Object empty() {
+                    return Optional.empty();
+                }
+                @Override
+                Object of(Object value) {
+                    return Optional.of(value);
+                }
```

Comment:
Do we need to cover this case? I think `Optional<T>` would already be handled by `OptionalCustomProvider` before `OptionalProvider` gets invoked.

Reply:
Right, it is not necessary